### PR TITLE
add `wrapattr`

### LINKED ISSUE #84: Add support for data streams (with known total size)
Hi,

It is easy to integrate tqdm with urllib when downloading a file, but when something is being uploaded, it is not so easy. One needs to intercept `.read()` or `.write()` calls and update tqdm manually.

I needed it in b2 command line tool, so I have [implemented it there](https://github.com/Backblaze/B2_Command_Line_Tool/pull/21/files), but maybe a stream wrapper could be useful for other tqdm users? This code is more specific to tqdm than to b2 command line tool. 

Do you want to support progress tracking of data streams?

### PR DESCRIPTION
**Pull request by casperdcl**
```python
# basic use
with tqdm.wrapattr(file_obj, "read", total=file_obj.size) as fobj:
    while True:
        chunk = fobj.read(chunk_size)
        if not chunk:
            break

# more bar control
from tqdm.utils import CallbackIOWrapper
with tqdm(total=file_obj.size) as t:
    fobj = CallbackIOWrapper(t.update, file_obj, "read")
    while True:
        chunk = fobj.read(chunk_size)
        if not chunk:
            break
    t.reset()
    # ... continue to use `t` for something else
```

- [x] provide class-level `wrapattr` context manager for patching `.read()` and `.write()` methods of objects
- [x] add `bytes=True` similar to CLI (`unit="B", unit_scale=True, unit_divisor=1024` but not `miniters=1`)
- [ ] document
  + [x] inline
  + [ ] readme
- [x] add tests
- [x] replace self.method
```python
pbar = tqdm(total=file_obj.size)
with pbar.self_wrapattr(file_obj, "read") as fobj:
```
with class.method
```python
with tqdm.wrapattr(file_obj, "read", total=file_obj.size) as fobj:
```
- [x] perhaps avoid `setattr` on external objects (return a new object, `CallbackIOWrapper`)?
- [x] expose `CallbackIOWrapper`
- [x] patch `DummyTqdmFile` (related to #713)
- fixes #84

### CODE REVIEW

**Comment by Suor** [comment]
Why do you monkey-patch particular functions instead of wrapping the whole file-like object?

**Comment by casperdcl** [comment]
Yes @Suor I was thinking it's 
probably safer to return a new object instead of patching (which may have unexpected side effects)

**Comment by casperdcl** [comment]
I also don't like the fact that `self.wrapattr` needs to be renamed (e.g. `self.self_wrapattr`) to avoid clashing with `cls.wrapattr`

**Comment by Suor** [comment]
If you wrap then you can make it work both ways. You also won't need to say read or write.

**Comment by Suor** [comment]
Also, I don't think you should provide `close` param. People are free to use `closing()` context manager.

**Comment by casperdcl** [comment]
> If you wrap then you can make it work both ways. You also won't need to say read or write.

Not really - there's a difference between `data = read(); update(len(data))` and `write(data); update(len(data))`. Wrapping both ways will cause unintended side effects for files open in `rw` mode where seeking is happening but the user only wants to see `write` progress.

> People are free to use `closing()` context manager.

Not needed; `tqdm` itself can be used as a context manager.

However, I view this whole PR as just being syntactic sugar. In which case:

```python
with tqdm(total=file_obj.size) as pbar:
    with pbar.self_wrapattr(file_obj, "read") as fobj:
        ...
```

becomes:

```python
pbar = tqdm(total=file_obj.size)
with pbar.self_wrapattr(file_obj, "read", close=True) as fobj:
    ...  # notice 1 level less indentation
```

Overall though @Suor I'm also still not ecstatic about this PR, though.

**Comment by Suor** [comment]
> Not really - there's a difference between

You can wrap both by default, this will cover the 99% of use cases where you only read or only write. Provide a way to block one of them. Or alternatively you follow whichever started first and block the second, this will cover even more.

> Not needed; tqdm itself can be used as a context manager.

Not needed if you implement it, but you are duplicating existing functionality:
```python
with closing(tqdm.wrap_file(raw_fd, ...)) as fd:
    ...
# total might be infered from `raw_fd.size`.
```
And no need to add a level.

> Overall though @Suor I'm also still not ecstatic about this PR, though.

This should be abstracted away somehow. If you don't want to add it here we can add it to `dvc.progress`, leaving a possibility to mainline it later. Bonus: unlike `tqdm` we can make non backwards compatible changes there.

File: tqdm/std.py [hunk 1]
```diff
@@ -126,6 +127,38 @@ def create_th_lock(cls):
...
+
+    def __setattr__(self, name, value):
+        return setattr(self._wrapped, name, value)
+
+    def __init__(self, callback, stream, method="read"):
+        """
+        Wrap a given `file`-like object's `read()` or `write()` to report
+        lengths to the given `callback`
+        """
+        object.__setattr__(self, '_wrapped', stream)
```
Comment (casperdcl) [inline tqdm/std.py]: @altendky just checking - do you think we need to hassle `encoding` here as with `SimpleTextIOWrapper`. I don't think we do.

File: tqdm/std.py [hunk 2]
```diff
@@ -127,6 +127,38 @@ def create_th_lock(cls):
...
+        func = getattr(stream, method)
+        if method == "write":
+            @wraps(stream.write)
```
Comment (Suor) [inline tqdm/std.py]: I don't think you should use `@wraps` here. This is a separate file-like object with its own write method.

File: tqdm/tests/tests_tqdm.py [hunk 3]
```diff
@@ -1800,3 +1800,18 @@ def test_auto():
...
+
+
+@with_setup(pretest, posttest)
+def test_wrapattr():
+    """Test wrapping file-like objects"""
+    data = "a twenty-char string"
+
+    with closing(StringIO()) as our_file:
+        with closing(StringIO()) as writer:
+            with tqdm.wrapattr(writer, "write", file=our_file) as wrap:
```
Comment (Suor) [inline tqdm/tests/tests_tqdm.py]: Should it be called `.wrap_file()`?

File: tqdm/std.py [hunk 4]
```diff
@@ -1442,27 +1474,10 @@ def self_wrapattr(self, stream, method, close=True):
...
-                self.update(len(res))
-                return res
-        else:
-            raise TqdmKeyError("Can only wrap read/write methods")
-
         try:
-            setattr(stream, method, wrapped)
-            yield stream
+            wrapped = CallbackIOWrapper(self.update, stream, method)
+            yield wrapped
```
Comment (Suor) [inline tqdm/std.py]: You can move the context manager functionality into `CallbackIOWrapper` and you won't need two methods here.
```python
def __enter__(self):
    return self
```

**Comment by casperdcl** [comment]
@Suor your suggestion is the class-based method (`wrapattr`) which I've already implemented. We're talking about the underlying object-based method (`self_wrapattr`, which I'd like to rename). This underlying method ideally shouldn't be used much by end-users anyway. I'm tempted to just call it `_wrapattr` which fixes the name issue and also technically allows me to change it in future without breaking public API backward-compat.

**Comment by Suor** [comment]
@casperdcl I don't understand why do you need any second method at all.

File: tqdm/std.py (hunk 2 above)
Comment (casperdcl) [reply tqdm/std.py]: given the `__get/setattr__` on the underlying `_wrapped` object I think it's fair to use `@wraps` though.

**Comment by casperdcl** [comment]
The `@classmethod` is really useful for exposing to all descendants (e.g. `tqdm_notebook`)

**Comment by Suor** [comment]
But why do you need an instance one?

File: tqdm/std.py (hunk 4 above)
Comment (casperdcl) [reply tqdm/std.py]: given this some more thought - I think you're right; now that I've abstracted most of the logic to `CallbackIOWrapper` the instance method convenience function is not required (it takes the same number of lines of code to use). This also gets rid of the annoying `close=True`, name clash, and generally seems cleaner.

File: tqdm/tests/tests_tqdm.py (hunk 3 above)
Comment (casperdcl) [reply tqdm/tests/tests_tqdm.py]: I was going for

1. something which looks like the built-in naming syntax `getattr`, `setattr`
2. something which wraps generic objects (files, streams, etc) so perhaps `.wrap_obj` or `.wrap_stream`

**Comment by casperdcl** [comment]
@Suor I think I've wound up essentially implementing what you were suggesting :)

File: tqdm/std.py [hunk 5]
```diff
@@ -126,6 +127,38 @@ def create_th_lock(cls):
...
+    def __getattr__(self, name):
+        return getattr(self._wrapped, name)
+
+    def __setattr__(self, name, value):
+        return setattr(self._wrapped, name, value)
+
+    def __init__(self, callback, stream, method="read"):
+        """
+        Wrap a given `file`-like object's `read()` or `write()` to report
+        lengths to the given `callback`
```
Comment (altendky) [inline tqdm/std.py]: Should probably at least document that it doesn't work for generators or anything else that doesn't work with `len()`?

File: tqdm/std.py (hunk 1 above)
Comment (altendky) [reply tqdm/std.py]: At first glance, no.  The purpose here isn't to encode but simply to let you have a chance to see how much stuff is going through.

**Comment by GF-Huang** [comment]
`tqdm.wrapattr` did not close the file_obj? How to make it close it automatic?

```python
with requests.get(f'https://xxxxxxxxxxxx.com/{file_name}', stream=False) as response, \
     tqdm.wrapattr(open(f'data/{file_name}', 'wb'), 'write', total=int(response.headers['content-length'])) as f:
        for chunk in response.iter_content(chunk_size=4096):
            f.write(chunk)

```

**Comment by abitrolly** [comment]
@GF-Huang I guess better open a new issue.

**Pull request merged**